### PR TITLE
Impl Server List Protocol (client & server).

### PR DESCRIPTION
**Pull request by toqueteos**
As seen in #40.

I renamed `ping` module to `slp`, makes more sense now, it's not just ping after all. 

/cc @atheriel

### CODE REVIEW

**Comment by toqueteos** [comment]
Build should fail only because of tests not working.

A vanilla server is required and JSON decoding is closely related to #37.

**QUESTION:** How can we flag that test to not run by default? /cc @eddyb @fenhl

**Comment by toqueteos** [comment]
Repushing to fix std_misc feature warning.

File: src/proto/slp.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    pub name: String,
+    pub protocol: i32,
+}
+
+/// Response sent to clients as JSON.
+#[derive(Debug)]
+#[derive_serialize]
+#[derive_deserialize]
+pub struct Response {
+    pub description: Description,
```
Comment (fenhl) [inline src/proto/slp.rs]: Either here or in #37, whichever is merged later, we should fix this: the `Description` struct should be replaced by `ChatJson`.

**Comment by atheriel** [comment]
@toqueteos Just remove the `#[test]` for now, I think. Or add an `#[ignore]`.

**Comment by atheriel** [comment]
This looks like a great start to me, by the way.

File: src/proto/slp.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    // C->S: Ping packet
+    let start = time::get_time();
+    try!(Ping { time: start.sec }.write(stream));
+
+    // S->C: Pong packet
+    match try!(Packet::read(stream)) {
+        Pong(_) => {
+            let end = time::get_time();
+            let elapsed = end.sub(start).num_milliseconds();
+            println!("StatusPing elapsed={}", elapsed);
```
Comment (toqueteos) [inline src/proto/slp.rs]: Forgot to remove this printnln! It may be interesting if logging is set to DEBUG or TRACE levels maybe?

**Comment by toqueteos** [comment]
@atheriel IRC guys at #rust told me `#[cfg()]` let's us do this... I'm currently searching how to run cargo test to trigger that.

File: src/proto/slp.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    use super::*;
+
+    use std::io::prelude::*;
+    use std::net::TcpStream;
+
+    use packet::handshake::Handshake;
+    use packet::{PacketWrite, NextState};
+
+    #[test]
+    #[cfg(vanilla_server_required)]
```
Comment (toqueteos) [inline src/proto/slp.rs:165]: This should do, now I need to know how to run `cargo test` with this cfg enabled.

**Comment by toqueteos** [comment]
Rebased onto master, should get green by Travis.

**Comment by toqueteos** [comment]
Currently broken because of serde dependencies.

**Comment by toqueteos** [comment]
I rebased onto master and got rid of serde as dependency, using plain old rustc-serialize as `ChatJson` does.

Vanilla server test is behind `#[cfg(vanilla_server_required)]` which works the same way as `#[ignore]` because as of today there's no way to pass `--cfg vanilla_server_required` to `cargo test`, confirmed on #rust IRC.

I confirmed it works with a local vanilla server on Windows, any of you can check this works.

I also made `Response.description` field a String, we can always go `String -> Json` later, still waiting for #37 to get merged.

**Comment by toqueteos** [comment]
Final repush to get rid of all warnings related to `slp` module.

**Pull request merged**